### PR TITLE
remove `randn?` from `torch.testing` namespace

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61841 Deprecate `assert_allclose` in favor of `assert_close`
* **#61840 remove `randn?` from `torch.testing` namespace**

Redo of #60859.

Differential Revision: [D29871017](https://our.internmc.facebook.com/intern/diff/D29871017)

### CODE REVIEW

File: torch/testing/_deprecated.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+"""
+
+import functools
+import warnings
+from typing import Any, Callable
+
+import torch
+
+
+__all__ = ["rand", "randn"]
```
Comment (mruberry) [inline torch/testing/_deprecated.py:13]: This is a cool technique

**Comment by mruberry** [comment]
There is only one appearance of "from torch.testing import randn_like" in test_autograd.py and no appearances of "from torch.testing import rand_like" within Facebook. 

"torch.testing.randn_like" appears in test_autograd.py and gradcheck.py, while "torch.testing.rand_like" never appears.

In addition to these, "make_non_contiguous" is imported by only one project outside PyTorch. It is used in test_torch.py and gradcheck.py, though, so maybe that's a good candidate for a follow-up PR?

"FileCheck" is used by several projects within Facebook, so that will require a focused follow-up.

"assert_allclose" is used by several projects OUTSIDE of Facebook, so that will require a proper deprecation (see comments on that PR -- seems like that deprecation should wait for a stable API to use as an alternative?)

The remaining functions exposed here:

https://github.com/pytorch/pytorch/blob/db11619901e1bd941042af4c49c6e9da7b0d206d/torch/testing/_core.py#L14

are all dtype getters, I think. Some of these dtype getters are used by external projects. "floating_types" and "integral_types" are both used by torchvision, for example, and "get_all_device_types" and "get_all_math_dtypes" are used by additional projects outside PyTorch. I think, however, that we can move all these to deprecated if someone tries to import them from testing and create internal versions of these functions. For torchvision we can just copy the functions they're using, and the few external projects can update themselves with a simple list of dtypes. Updating our internal test suite to use the internal versions of these operations will take some doing, however. A lot of our test suites are calling torch.testing.get_all_device_types(), for example.

How does that sound, @pmeier?

**Comment by pmeier** [comment]
> There is only one appearance of "from torch.testing import randn_like" in test_autograd.py and no appearances of "from torch.testing import rand_like" within Facebook.
>
> "torch.testing.randn_like" appears in test_autograd.py and gradcheck.py, while "torch.testing.rand_like" never appears.

Seems like this PR is good to go then.

> In addition to these, "make_non_contiguous" is imported by only one project outside PyTorch. It is used in test_torch.py and gradcheck.py, though, so maybe that's a good candidate for a follow-up PR?

Yes, I'll do that one next in the stack.

> "FileCheck" is used by several projects within Facebook, so that will require a focused follow-up.

Sounds good.

> "assert_allclose" is used by several projects OUTSIDE of Facebook, so that will require a proper deprecation (see comments on that PR -- seems like that deprecation should wait for a stable API to use as an alternative?)

For reference: #61844. 

I'm against waiting that long for a deprecation for two reasons:

1. It can easily be confused with `assert_close` in the same namespace. Since `assert_close` will be the successor I think we should avoid that.
2. I would relax the "only deprecate in favor of stable" towards "only deprecate in favor of something more stable". If we assume stable > beta > prototype > undocumented we would be good to deprecate right away. 

> The remaining functions exposed here:
> 
> https://github.com/pytorch/pytorch/blob/db11619901e1bd941042af4c49c6e9da7b0d206d/torch/testing/_core.py#L14
> 
> are all dtype getters, I think. Some of these dtype getters are used by external projects. "floating_types" and "integral_types" are both used by torchvision, for example, and "get_all_device_types" and "get_all_math_dtypes" are used by additional projects outside PyTorch. I think, however, that we can move all these to deprecated if someone tries to import them from testing and create internal versions of these functions. For torchvision we can just copy the functions they're using, and the few external projects can update themselves with a simple list of dtypes. 

Agreed. The largest problem that I see is that the naming scheme is very confusing if one is not aware of the organic growth of dtypes over time. For example, I bet other libraries using `get_all_math_dtypes` do not know that they are not in fact getting all math dtypes:

https://github.com/pytorch/pytorch/blob/1ce3281a6d266d627f75a283dac5ce10c9f4a27b/torch/testing/_core.py#L386-L388

> Updating our internal test suite to use the internal versions of these operations will take some doing, however. A lot of our test suites are calling torch.testing.get_all_device_types(), for example.

Should be a fairly easy substitution. Might be some busy work, but I don't expect any larger obstacles.

**Review by mruberry — APPROVED** [review]
Cool approach to deprecations

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29871017).

**Pull request closed without merging**